### PR TITLE
password_expire support for mysql_user

### PR DESCRIPTION
**Pull request by tompal3**
##### SUMMARY

Add new feature to support password expire policy for users as discussed in: https://github.com/ansible-collections/community.mysql/issues/513

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

- mysql_user
##### ADDITIONAL INFORMATION

 new features will include setting password_lifetime policy for creating  or altering user: 
 - expire password in custom number of days.
 - expire password using global variable `default_password_lifetime`.
 - expire password now.
 - never expire password.

```paste below

```

### CODE REVIEW

**Comment by laurent-indermuehle** [comment]
Hi @tompal3 . Great start! Thanks for your contribution.

When I started to contribute to c.mysql I didn't knew how to find the error messages from the failing tests. So I paste it here in case you didn't saw it already:

```
TASK [test_mysql_role : Create role0 in check_mode again] **********************
fatal: [testhost]: FAILED! => {"changed": false, "msg": "user_mod() missing 2 required positional arguments: 'password_expire' and 'password_expire_interval'"}
```

To find it, I searched for "FAILED" and hit enter until I saw an error without the light blue text "...ignoring"

Do you know why this is failing?

**Comment by tompal3** [comment]
Thanks for the tip @laurent-indermuehle. Yes I have already fixed it, but now I'm trying to setup ansible integration tests on my wsl ubuntu instance to reduce number of commits and to write some tests for my PR.

**Comment by laurent-indermuehle** [comment]
In that case you'll need to adapt the Makefile that I wrote for podman. I think docker will work best in your case. Feel free to ask questions here or in matrix #mysql:ansible.com if you need help.

**Comment by laurent-indermuehle** [comment]
@tompal3 as you mention on Matrix, our scheduled integration tests for the plugins fails with latest Ansible due to jinja templating been forbidden in the assertions. I'm fixing this now. You'll be able to rebase your fork on this to avoid clutter in your patch when I'm done. Thanks for your message.

File: plugins/modules/mysql_user.py [hunk 1]
```diff
@@ -155,6 +155,18 @@
       - Cannot be used to set global variables, use the M(community.mysql.mysql_variables) module instead.
     type: dict
     version_added: '3.6.0'
+  password_expire:
+    description:
+      - C(never) password will never expire.
+      - C(default) password is defined ussing global system varaiable I(default_password_lifetime) setting.
```
Comment (laurent-indermuehle) [inline plugins/modules/mysql_user.py]:
```suggestion
      - C(default) password is defined using global system variable I(default_password_lifetime) setting.
```

File: plugins/modules/mysql_user.py [hunk 2]
```diff
@@ -461,6 +477,15 @@ def main():
...
         module.fail_json(msg=mysql_driver_fail_msg)
 
+    if password_expire == "interval" and not password_expire_interval:
+        module.fail_json(msg="password_expire value interval \
+                         should be used with password_expire_interval")
+
+    if password_expire_interval:
+        if password_expire_interval < 1:
+            module.fail_json(msg="password_expire_interval value \
+                             should be positive number")
```
Comment (laurent-indermuehle) [inline plugins/modules/mysql_user.py]:
```suggestion
    if password_expire_interval and password_expire_interval < 1:
        module.fail_json(msg="password_expire_interval value \
                             should be positive number")
```

File: tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        password_expire_interval: 20
+        state: present
+      register: result
+      check_mode: True
+      ignore_errors: true
+
+    - name: password_expire | Assert that check mode for user creation 
+      assert:
+        that:
+          - result.changed == true
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml]:
```suggestion
      failed_when: result is not changed
```

File: tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        - username: "{{ user_name_3 }}"
+          password_expire: interval
+          password_expire_interval: 10
+          check_mode: true
+          expect_change: true
+          expected_password_lifetime: "0"
+          expected_password_expired: "N"
+
+    - name: password_expire | Set password_expire = interval without password_expire_interval
+      mysql_user:
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml]:
```suggestion
      community.mysql.mysql_user:
```

File: tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        name: '{{ user_name_4 }}'
+        host: '%'
+        password: '{{ user_password_4 }}'
+        password_expire: interval
+        state: present
+      register: result
+      ignore_errors: true
+
+    - name: password_expire | Assert that action fails if 'password_expire_interval' not set
+      assert:
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml]:
```suggestion
      ansible.builtin.assert:
```

File: tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        that:
+          - result is failed
+          - "'should be used with' in result.msg"
+
+    - name: password_expire | Set password_expire_interval < 1
+      mysql_user:
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml]:
```suggestion
      community.mysql.mysql_user:
```

File: tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        host: '%'
+        password: '{{ user_password_4 }}'
+        password_expire: interval
+        password_expire_interval: -1
+        state: present
+      register: result
+      ignore_errors: true
+
+    - name: password_expire | Assert that action fails if 'password_expire_interval' is < 1
+      assert:
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml]:
```suggestion
      ansible.builtin.assert:
```

File: tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        that:
+          - result is failed
+          - "'should be positive number' in result.msg"
+
+    - name: password_expire | check mode for user creation
+      mysql_user:
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml]:
```suggestion
      community.mysql.mysql_user:
```

File: tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+        password: '{{ user_password_4 }}'
+        password_expire: interval
+        password_expire_interval: 20
+        state: present
+      register: result
+      check_mode: True
+      ignore_errors: true
+
+    - name: password_expire | Assert that check mode for user creation 
+      assert:
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/test_password_expire.yml]:
```suggestion
      ansible.builtin.assert:
```

File: tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml [hunk 10]
```diff
@@ -0,0 +1,74 @@
+---
+- name: Utils | Assert user password_expire | Create modify {{ username }} with password_expire
+  mysql_user:
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml]:
```suggestion
  community.mysql.mysql_user:
```

File: tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+  check_mode: "{{ check_mode | default(false) }}"
+
+- name: Utils | Assert user password_expire | Assert a change occurred
+  assert:
+    that: result.changed  ==  expect_change_value
+  vars:
+    expect_change_value: "{{ expect_change }}"
+
+- name: Utils | Assert user password_lifetime | Query user '{{ username }}'
+  command: '{{ mysql_command }} -BNe "SELECT IFNULL(password_lifetime, -1) FROM mysql.user where user=''{{ username }}'' and host=''{{ host }}''"'
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml]:
```suggestion
  ansible.builtin.command:
    cmd: >
      {{ mysql_command }} -BNe "SELECT IFNULL(password_lifetime, -1)
      FROM mysql.user where user='{{ username }}' and host='{{ host }}'"
```

File: tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+    that:
+      - expected_password_lifetime_value in password_lifetime.stdout_lines
+  vars:
+    expected_password_lifetime_value: "{{ expected_password_lifetime }}"
+  when:
+    - db_engine == 'mysql'
+    - db_version is version('5.7.0', '>=')
+
+- name: Utils | Assert user password_lifetime | Query user '{{ username }}'
+  command:
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml]:
```suggestion
  ansible.builtin.command:
```

File: tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+      - expected_password_lifetime_value in password_lifetime.stdout_lines
+  vars:
+    expected_password_lifetime_value: "{{ expected_password_lifetime }}"
+  when:
+    - db_engine == 'mariadb'
+    - db_version is version('10.4.3', '>=')
+
+- name: Utils | Assert user password_expired | Query user '{{ username }}'
+  command: "{{ mysql_command }} -BNe \"SELECT password_expired FROM mysql.user \
+    WHERE user='{{ username }}' and host='{{ host }}'\""
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml]:
```suggestion
  ansible.builtin.command:
    cmd: >
      {{ mysql_command }} -BNe "SELECT password_expired FROM mysql.user 
      WHERE user='{{ username }}' and host='{{ host }}'"
```

File: tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+    (db_engine == 'mariadb' and db_version is version('10.4.3', '>='))
+
+- name: Utils | Assert user password_expired | Assert password_expired is in user stdout
+  assert:
+    that:
+      - expected_password_expired_value in password_expired.stdout_lines
+  vars:
+    expected_password_expired_value: "{{ expected_password_expired }}"
+  when: (db_engine == 'mysql' and db_version is version('5.7.0', '>=')) or
+    (db_engine == 'mariadb' and db_version is version('10.4.3', '>='))
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml]:
```suggestion
  failed_when: expected_password_expired_value not in password_expired.stdout_lines
  vars:
    expected_password_expired_value: "{{ expected_password_expired }}"
```

File: tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+- name: Utils | Assert user password_lifetime | Assert password_lifetime is in user stdout
+  assert:
+    that:
+      - expected_password_lifetime_value in password_lifetime.stdout_lines
+  vars:
+    expected_password_lifetime_value: "{{ expected_password_lifetime }}"
+  when:
+    - db_engine == 'mariadb'
+    - db_version is version('10.4.3', '>=')
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml]:
```suggestion
  failed_when: expected_password_lifetime_value not in password_lifetime.stdout_lines
  vars:
    expected_password_lifetime_value: "{{ expected_password_lifetime }}"
```

File: tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+
+- name: Utils | Assert user password_lifetime | Assert password_lifetime is in user stdout
+  assert:
+    that:
+      - expected_password_lifetime_value in password_lifetime.stdout_lines
+  vars:
+    expected_password_lifetime_value: "{{ expected_password_lifetime }}"
+  when:
+    - db_engine == 'mysql'
+    - db_version is version('5.7.0', '>=')
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml]:
```suggestion
  failed_when: expected_password_lifetime_value not in password_lifetime.stdout_lines
  vars:
    expected_password_lifetime_value: "{{ expected_password_lifetime }}"
```

File: tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    password_expire: "{{ password_expire }}"
+    password_expire_interval: "{{ password_expire_interval | default(omit) }}"
+  register: result
+  check_mode: "{{ check_mode | default(false) }}"
+
+- name: Utils | Assert user password_expire | Assert a change occurred
+  assert:
+    that: result.changed  ==  expect_change_value
+  vars:
+    expect_change_value: "{{ expect_change }}"
```
Comment (laurent-indermuehle) [inline tests/integration/targets/test_mysql_user/tasks/utils/assert_user_password_expire.yml]:
```suggestion
  failed_when: result.changed != expect_change_value
  vars:
    expect_change_value: "{{ expect_change }}"
```

**Review by laurent-indermuehle — CHANGES_REQUESTED** [review]
I'm quite rusty on the Python part. Also I may have introduced some regression so I apologise in advance if that happen.

As a rule of thumb, if an assertion has only one test, I prefer to use failed_when. This offer a cleaner output (searching for a failed task is easier when you don't have thousand of "failed... ignored" in the log.

**Comment by tompal3** [comment]
> I'm quite rusty on the Python part. Also I may have introduced some regression so I apologise in advance if that happen.
> 
> As a rule of thumb, if an assertion has only one test, I prefer to use failed_when. This offer a cleaner output (searching for a failed task is easier when you don't have thousand of "failed... ignored" in the log.

Thanks for the code review, committed all your suggestions.

File: plugins/modules/mysql_user.py [hunk 18]
```diff
@@ -155,6 +155,18 @@
...
     type: dict
     version_added: '3.6.0'
+  password_expire:
+    description:
+      - C(never) password will never expire.
+      - C(default) password is defined using global system variable I(default_password_lifetime) setting.
+      - C(interval) password will expire in days which is defined in I(password_expire_interval).
+      - C(now) password will expire immediately.
+    type: str
+    choices: [ now, never, default, interval ]
```
Comment (Andersson007) [inline plugins/modules/mysql_user.py:165]:
```suggestion
    choices: [ now, never, default, interval ]
    version_added: '3.9.0'
```

File: plugins/modules/mysql_user.py [hunk 19]
```diff
@@ -155,6 +155,18 @@
...
+  password_expire_interval:
+    description:
+      - number of days password will expire. Used when I(password_expire=interval).
+    type: int
```
Comment (Andersson007) [inline plugins/modules/mysql_user.py:170]:
```suggestion
    type: int
    version_added: '3.9.0'
```

File: plugins/modules/mysql_user.py [hunk 20]
```diff
@@ -155,6 +155,18 @@
...
+    description:
+      - C(never) password will never expire.
+      - C(default) password is defined using global system variable I(default_password_lifetime) setting.
+      - C(interval) password will expire in days which is defined in I(password_expire_interval).
+      - C(now) password will expire immediately.
+    type: str
+    choices: [ now, never, default, interval ]
+  password_expire_interval:
+    description:
+      - number of days password will expire. Used when I(password_expire=interval).
```
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
      - Number of days I(password) will expire. Requires I(password_expire=interval).
```

File: plugins/modules/mysql_user.py [hunk 21]
```diff
@@ -415,6 +427,8 @@ def main():
         force_context=dict(type='bool', default=False),
         session_vars=dict(type='dict'),
         column_case_sensitive=dict(type='bool', default=None),  # TODO 4.0.0 add default=True
+        password_expire=dict(type='str', choices=['now', 'never', 'default', 'interval'], no_log=True),
+        password_expire_interval=dict(type='int', no_log=True),
```
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
        password_expire=dict(type='str', choices=['now', 'never', 'default', 'interval'], no_log=True),
        password_expire_interval=dict(type='int', no_log=True),
```
why do we `no_log` them?

File: plugins/modules/mysql_user.py [hunk 22]
```diff
@@ -461,6 +477,14 @@ def main():
     if mysql_driver is None:
         module.fail_json(msg=mysql_driver_fail_msg)
 
+    if password_expire == "interval" and not password_expire_interval:
+        module.fail_json(msg="password_expire value interval \
+                         should be used with password_expire_interval")
+
+    if password_expire_interval and password_expire_interval < 1:
+        module.fail_json(msg="password_expire_interval value \
+                             should be positive number")
```
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
    if password_expire == "interval" and not password_expire_interval:
        module.fail_json(msg="password_expire value interval \
                         should be used with password_expire_interval")

    if password_expire_interval and password_expire_interval < 1:
        module.fail_json(msg="password_expire_interval value \
                             should be positive number")
```
Can we use one of keys in AnsibleModue initialization instead? like `required_if` / `required_by`, etc.? See the keys at the bottom of the [doc section](https://docs.ansible.com/ansible/latest/dev_guide/developing_program_flow_modules.html#argument-spec).
Could you please check? if not possible, it's OK as is.

File: plugins/modules/mysql_user.py [hunk 23]
```diff
@@ -155,6 +155,18 @@
       - Cannot be used to set global variables, use the M(community.mysql.mysql_variables) module instead.
     type: dict
     version_added: '3.6.0'
+  password_expire:
+    description:
+      - C(never) password will never expire.
+      - C(default) password is defined using global system variable I(default_password_lifetime) setting.
+      - C(interval) password will expire in days which is defined in I(password_expire_interval).
+      - C(now) password will expire immediately.
```
Comment (Andersson007) [inline plugins/modules/mysql_user.py]:
```suggestion
      - C(never) - I(password) will never expire.
      - C(default) - I(password) is defined using global system variable I(default_password_lifetime) setting.
      - C(interval) - I(password) will expire in days which is defined in I(password_expire_interval).
      - C(now) - I(password) will expire immediately.
```

File: plugins/module_utils/user.py [hunk 24]
```diff
@@ -924,6 +956,77 @@ def limit_resources(module, cursor, user, host, resource_limits, check_mode):
...
+        statment = "PASSWORD EXPIRE INTERVAL %d DAY" % (password_expire_interval)
+    elif password_expire.lower() == "now":
+        statment = "PASSWORD EXPIRE"
+
+    params = (user, host)
+    query = ["ALTER USER %s@%s"]
+
+    query.append(statment)
+    query = ' '.join(query)
+    cursor.execute(query, params)
```
Comment (Andersson007) [inline plugins/module_utils/user.py]:
```suggestion
    cursor.execute("ALTER USER %s@%s " + statement, (user, host))
```

File: plugins/module_utils/user.py [hunk 25]
```diff
@@ -924,6 +956,77 @@ def limit_resources(module, cursor, user, host, resource_limits, check_mode):
...
+        password_expire_days (int): Invterval of days password expires.
+    """
+    if password_expire.lower() == "never":
+        statment = "PASSWORD EXPIRE NEVER"
+    elif password_expire.lower() == "default":
+        statment = "PASSWORD EXPIRE DEFAULT"
+    elif password_expire.lower() == "interval":
+        statment = "PASSWORD EXPIRE INTERVAL %d DAY" % (password_expire_interval)
+    elif password_expire.lower() == "now":
+        statment = "PASSWORD EXPIRE"
```
Comment (Andersson007) [inline plugins/module_utils/user.py]:
```suggestion
    if password_expire.lower() == "never":
        statement = "PASSWORD EXPIRE NEVER"
    elif password_expire.lower() == "default":
        statement = "PASSWORD EXPIRE DEFAULT"
    elif password_expire.lower() == "interval":
        statement = "PASSWORD EXPIRE INTERVAL %d DAY" % (password_expire_interval)
    elif password_expire.lower() == "now":
        statement = "PASSWORD EXPIRE"
```

File: plugins/module_utils/user.py [hunk 26]
```diff
@@ -301,6 +309,30 @@ def user_mod(cursor, user, host, host_all, password, encrypted,
...
+                    update = True
+
+                    if module.check_mode:
+                        return {'changed': True, 'msg': msg, 'password_changed': password_changed}
+                    set_password_expire(cursor, user, host, password_expire, password_expire_interval)
+                    password_changed = True
+                    changed = True
+            else:
+                module.fail_json(msg="The server version does not match the requirements "
+                                 "for password_expire parameter. See module's documentation.")
```
Comment (Andersson007) [inline plugins/module_utils/user.py]:
```suggestion
            if not impl.server_supports_password_expire(cursor):
                module.fail_json(msg="The server version does not match the requirements "
                                     "for password_expire parameter. See module's documentation.")
            update = False
            mariadb_role = True if "mariadb" in str(impl.__name__) else False
            current_password_policy = get_password_expiration_policy(cursor, user, host, maria_role=mariadb_role)
            password_expired = is_password_expired(cursor, user, host)
            # Check if changes needed to be applied.
            if not ((current_password_policy == -1 and password_expire == "default") or
                    (current_password_policy == 0 and password_expire == "never") or
                    (current_password_policy == password_expire_interval and password_expire == "interval") or
                    (password_expire == 'now' and password_expired)):

                update = True

                if module.check_mode:
                    return {'changed': True, 'msg': msg, 'password_changed': password_changed}
                set_password_expire(cursor, user, host, password_expire, password_expire_interval)
                password_changed = True
                changed = True               
```
the only changed part is:
```
if not impl.server_supports_password_expire(cursor):
                module.fail_json(msg="The server version does not match the requirements "
                                     "for password_expire parameter. See module's documentation.")
```
in the rest, the indentation level is the only changed thing

**Review by Andersson007 — COMMENTED** [review]
@tompal3 hi, thanks for the great improvement!
Could you please add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#creating-a-changelog-fragment) to let users know about the added feature?
Any questions? Just ask
Thanks!

**Comment by Andersson007** [comment]
hello, any updates on this?

**Comment by tompal3** [comment]
> hello, any updates on this?

Hi, sorry missed your previous comment, will check it this week.

File: plugins/modules/mysql_user.py (hunk 21 above)
Comment (tompal3) [reply plugins/modules/mysql_user.py]: no reason why we should, will remove that one.

File: plugins/modules/mysql_user.py (hunk 22 above)
Comment (tompal3) [reply plugins/modules/mysql_user.py]: can only move if statement of `password_expire`: `interval` is used together with `password_expire_interval`. But can't check `password_expire_interval` to be positive integer in `argument_spec`.

Comment (Andersson007) [reply plugins/modules/mysql_user.py]: @tompal3 

>can only move if statement of password_expire: interval is used together with password_expire_interval.

It'd be fine, could you please do it for this one?

**Comment by Andersson007** [comment]
@tompal3 could you also please [rebase](https://docs.ansible.com/ansible/latest/dev_guide/developing_rebasing.html) the PR, fix the conflicts, and force-push?

**Comment by Andersson007** [comment]
@tompal3 there's a sanity failure:
```
ERROR: Found 1 pep8 issue(s) which need to be resolved:
ERROR: plugins/module_utils/user.py:1071:1: E302: expected 2 blank lines, found 1 (50%)
```
needs to be fixed before merging

**Comment by tompal3** [comment]
> @tompal3 there's a sanity failure:
> 
> ```
> ERROR: Found 1 pep8 issue(s) which need to be resolved:
> ERROR: plugins/module_utils/user.py:1071:1: E302: expected 2 blank lines, found 1 (50%)
> ```
> 
> needs to be fixed before merging

@Andersson007 ok, I have remembered now, sanity checks fails if variable containing *password* key word  is not set to `no_log=True`. So I will leave that one there.

**Review by laurent-indermuehle — APPROVED** [review]

**Comment by laurent-indermuehle** [comment]
Only way I found to get rid of an old request for changes I did was to validate the whole PR.  Thought I'm too busy now to do a full review, so please ignore my vote on this one (I skimmed, it looks not bad at all thought). You can ping me if necessary.

File: changelogs/fragments/598-password_expire-support-for-mysql_user.yml [hunk 27]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - "mysql_user - new feature - password expiration management for mysql user (https://github.com/ansible-collections/community.mysql/pull/598)."
```
Comment (Andersson007) [inline changelogs/fragments/598-password_expire-support-for-mysql_user.yml]:
```suggestion
  - "mysql_user - add the ``password_expire`` and ``password_expire_interval`` arguments to implement the password expiration management for mysql user (https://github.com/ansible-collections/community.mysql/pull/598)."
```

Comment (Andersson007) [reply changelogs/fragments/598-password_expire-support-for-mysql_user.yml]: @tompal3 the last one, then we'll merge, thanks

**Comment by Andersson007** [comment]
@tompal3 thanks for the contributing!
@laurent-indermuehle thanks for the review!

@tompal3 FYI we have a forum group and a matrix channel, see the communication guide for [details](https://docs.ansible.com/ansible/devel/community/communication.html).

**Pull request merged**